File: typescript/src/main/vsdk/toolkit/skin/Compound.ts
import { Ray } from "../common/linealAlgebra/Ray";
import { Statistics } from "../common/statistics/Statistics";
import { Geometry } from "./Geometry";
import { GeometryClassId } from "./GeometryClassId";
import type { RayHit } from "./RayHit";

export class Compound extends Geometry {
  public children: Geometry[] | null;

  public constructor(geometryList: Geometry[] | null) {
    super(GeometryClassId.COMPOUND);
    Statistics.instance().reader.numberOfCompounds++;
    this.children = geometryList;
    Geometry.listBounds(this.children, this.boundingBox);
    this.boundingBox.enlargeTinyBit();
    this.bounded = true;
  }

  public override destroy(): void {
    Statistics.instance().reader.numberOfCompounds--;
    if (this.children !== null) {
      this.children = null;
    }
    super.destroy();
  }

  public override discretizationIntersect(
    ray: Ray,
    minimumDistance: number,
    maximumDistance: number[],
    hitFlags: number,
    hitStore: RayHit | null
  ): RayHit | null {
    if (!this.discretizationIntersectPreTest(ray, minimumDistance, maximumDistance)) {
      return null;
    }

    return Geometry.listDiscretizationIntersect(
      this.children,
      ray,
      minimumDistance,
      maximumDistance,
      hitFlags,
      hitStore
    );
  }
}

File: java/src/main/java/vsdk/toolkit/skin/Compound.java
package vsdk.toolkit.skin;

import vsdk.toolkit.environment.geometry.elements.*;

import java.util.ArrayList;
import vsdk.toolkit.common.linealAlgebra.Ray;
import vsdk.toolkit.common.statistics.Statistics;

public class Compound extends Geometry {
    public ArrayList<Geometry> children;

    /**
    Creates a Compound from a list of geometries.

    Actually, it just counts the number of compounds in the scene and
    returns the geometry list.
    */
    public Compound(ArrayList<Geometry> geometryList) {
        super(GeometryClassId.COMPOUND);
        Statistics.instance().reader.numberOfCompounds++;
        children = geometryList;
        Geometry.listBounds(children, boundingBox);
        boundingBox.enlargeTinyBit();
        bounded = true;
    }

    @Override
    public void destroy() {
        Statistics.instance().reader.numberOfCompounds--;
        if (children != null) {
            children = null;
        }
        super.destroy();
    }

    /**
    DiscretizationIntersect returns null if the ray does not hit the discretization
    of the object. If the ray hits the object, a hit record is returned containing
    information about the intersection point. See geometry.h for more explanation.
    */
    @Override
    public RayHit discretizationIntersect(
        Ray ray,
        float minimumDistance,
        float[] maximumDistance,
        int hitFlags,
        RayHit hitStore) {
        if (!discretizationIntersectPreTest(ray, minimumDistance, maximumDistance)) {
            return null;
        }

        return Geometry.listDiscretizationIntersect(children, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
    }
}
